### ISSUE TITLE
allow no routes to be defined in typespec-csharp

### ISSUE THREAD

**Issue by cataggar** [Azure/autorest.csharp#3423]
Related to the proposed `generate-all-models: true` from #2960 #2974, it would be great to not error when no routes are defined. I have several places that I want to use `generate-all-models: true` and not define any routes. Currently, you get an error with `"@azure-tools/typespec-csharp": "0.2.0-beta.20230518.1"` of:

```
PS C:\Users\cataggar\ms\avs\src\conveyor\ConsumptionWorker\client> tsp compile .
TypeSpec compiler v0.44.0

{"level":"info","message":"routes:0"}
<unknown location>:1:1 - warning @typespec/http/no-routes: Current spec is not exposing any routes. This could be to not having the service namespace marked with @service.
> 1 |
    | ^
error @azure-tools/typespec-csharp/No-Route: No Route for service for service Pet
```

Using this main.tsp:
```
import "@typespec/http";
import "@typespec/versioning";

using TypeSpec.Http;
using TypeSpec.Versioning;

@service({
  title: "Pet API",
})
@versioned(ServiceApiVersions)
namespace Pet;

enum ServiceApiVersions {
  v2023_05: "2023-05",
}

@discriminator("kind")
model Pet {
  // kind: string;
}

model Cat extends Pet {
  kind: "cat";
  meow: boolean;
}

model Dog extends Pet {
  kind: "dog";
  bark: boolean;
}

// @route("/pet")
// @get
// op pet(): Pet;
```

**Comment by chunyu3** [comment]
Hello @cataggar Do you want to generate models which is not used in any operation? We proposed this `Generate-All-Models` feature because we cannot get the right Usage of this models, and if it changes later such as add a operation to use some model, the model's usage will change which will cause breaking change. 
And csharp codegen will generate internal serialize/deserialize method for this model which cannot be used outside.

@cataggar  could you help to provide some context about  what scenario do we need to generate the model? Seems such model will never be used. We will discuss with our arch about it. Thanks

**Comment by cataggar** [comment]
Yes, I want to use the generated types & serialization/deserialization from the model definitions. My current workaround is to define fake http routes. We want to eventually model our Service Bus API specs in TypeSpec, but just the models is a good starting point.

**Comment by chunyu3** [comment]
Hello @cataggar Current our serialization/deserialization methods are internal, and it cannot be used outside.  Can you tell me how do you use them? move them to customized code, and change it `public`?

**Comment by chunyu3** [comment]
> We are interested in moving from Swagger to TypeSpec. In this [internal PR](https://dev.azure.com/msazure/One/_git/Azure-Dedicated-AVS/pullrequest/8050535), I'm trying it out for one small internal API. The AutoRest setting we were using was `public-clients: true`.

Hello @cataggar If the model is not used by public method, it will be `internal`, and if the model is not used, it will be removed. You can set the emitter option `unreferenced-types-handling: keepAll` to let the model's accessibility `public`.

**Comment by l0lawrence** [comment]
Hello,

Adding on to this issue - we also want to generate a models-only package for [EventGrid SystemEvents](https://github.com/l0lawrence/azure-rest-api-specs/blob/system_events_temp/specification/eventgrid/Azure.Messaging.EventGrid/client_systemEvents.tsp). 

We have decorated the models with @access `public` and @usage `input | output` , but we run into this error when trying to generate with the emitter.

**Comment by cataggar** [comment]
`keepAll` is working in combination with using `@access(Access.public) and `@usage(Usage.input | Usage.output)`.

```
@access(Access.public)
@usage(Usage.input | Usage.output)
model AuditApplication {
    resultDescription?: string;
    AuditCategories?: string[];
    OperationName?: string;
    order_info?: string;
}
```

**Comment by cataggar** [comment]
Sorry, that is for code generation for models not in routes. That is working. I'm not sure about no routes. I'll try that next month.
